### ISSUE TITLE
Chipper and perennial have copies of execute.js, and one is stale

### ISSUE THREAD

**Issue by samreid** [phetsims/chipper#1018]
Assigning the developers with the most commits on those files.  How do you recommend to proceed?

**Comment by jonathanolson** [comment]
They are quite different, presumably best to leave separate?

**Comment by zepumph** [comment]
>They are quite different, presumably best to leave separate?



Yes but not for any good reason. It just looks like they have grown as they have needed to, but I see no reason, besides the annoyance of this problem, that stderr as an arg to ExecuteError (in perennial but not chipper), or `shell` option (in chipper but not perennial) shouldn't be supported generally.

The only issue I see is winston because it is used by the build server. Perhaps we can still work around that.

**Comment by jonathanolson** [comment]
> The only issue I see is winston because it is used by the build server. Perhaps we can still work around that.

And chipper prefers grunt logging? Could we wrap chipper to use winston? I definitely have a preference for the perennial version in general (with the `shell` option added?)

**Comment by samreid** [comment]
Should we adopt the same policy as SimVersion, which has the primary copy in perennial and one copied to chipper on a regular basis?

**Comment by jonathanolson** [comment]
> Should we adopt the same policy as SimVersion, which has the primary copy in perennial and one copied to chipper on a regular basis?

Maybe there should be a directory of code which works like this?

**Comment by zepumph** [comment]
Good idea!

**Comment by samreid** [comment]
I'm not sure of the best directory location or name.  How about:

perennial/js/chipper/ ?

**Comment by samreid** [comment]
I recommend to have the destination directory in chipper be named using a similar pattern.  Perhaps the copy will be from `perennial/js/chipper/` to `chipper/js/perennial/`

**Comment by samreid** [comment]
On second thought, those names seem too awkward.  It sounds like perennial/js/chipper code is intended primarily for chipper, which is not the case. Perhaps we should create a better name to unify these.  A unique name (not used widely in our project) to mean "used in both perennial and chipper".  Perhaps "biannual" or "dual".  Or maybe perennial would have dual-source and chipper would have dual-target.  Or we could use parent/child or original/copy.  Maybe we'll start with dual and see how that goes.

**Comment by samreid** [comment]
I moved SimVersion to dual/ and ran `grunt update` on all active repos.  I tested the following things:

Launching an unbuilt simulation
Building a simulation, launching it

I could not test a chains rc release with working copy changes, so I'll commit next.  This is going to span many repos since a preload path changed.

**Comment by samreid** [comment]
All changes pushed, and I confirmed a chains phet-io/phet RC completed successfully.  Next this issue should move execute.js to dual/

**Comment by samreid** [comment]
Putting this issue on hold until we discuss disadvantages of this approach and alternatives as part of https://github.com/phetsims/phet-io/issues/1733#issuecomment-877301298.

**Comment by zepumph** [comment]
I think that execute is a much more tolerable file to copy to chipper than the suite of phet-io api generation files. Furthermore I think that there are a selection of functions (mostly utility/tools) like `withServer` over in https://github.com/phetsims/perennial/issues/175 that I can't see another way to support using in chipper and perennial.

**Comment by samreid** [comment]
@zepumph do you want a `git add` command to run automatically, or should it be manual for now?

**Comment by samreid** [comment]
OK I moved execute to dual/ and tested a few functions in perennial and chipper.  I don't feel like I have tested everything, but I don't know how to do that.  I think the next step is a code review.  Other improvements will be on their way as part of https://github.com/phetsims/perennial/issues/231.

@zepumph or @jonathanolson can you please review?

**Comment by zepumph** [comment]
Looking really nice. There is a new issue in which update-copyright-dates isn't run in perennial (since it is a chipper task). So daily grunt work is going to fight itself every night. Want to move that script to perennial?

**Comment by samreid** [comment]
I'm not opposed to updating the copyright dates in perennial, but it seems this proposal entails moving the following files to perennial/dual/.

* updateCopyrightDate.js
* updateCopyrightDates.js
* updateCopyrightDatesForGeneratedFile.js
* getCopyrightLine.js

Does the daily grunt work fighting itself mean: 
(a) bad copyright dates in perennial/dual
(b) file gets copied to chipper/dual with bad copyright date
(c) copyright date is corrected in dual by daily-grunt-work before commit?

Since we don't have any short-circuits anyways (like "don't copy the file if it has the same contents") it seems a lower priority fight?  Just want to make sure I understand the problem here.  I'm not a fan of putting too many things in perennial/dual, but my life did get better once I realized I could tell my IDE to exclude chipper/dual.

**Comment by zepumph** [comment]
> I realized I could tell my IDE to exclude chipper/dual.

And I think we all should!

It is fighting because first in the daily grunt work task, copyright dates are updated, which means that chipper/dual/ files get updated to appropriate copyright dates, but perennial files don't. Then later in the task perennial files get copied and replace the new copyright dates in chipper/dual/. Thus, every day there will be a useless commit on files in chipper/dual. This is annoying for git history. We could also just manually update files that are in dual, right now, but that won't work long term.

**Comment by zepumph** [comment]
My general preference is to go all in on `dual/`, and when we hate reevaluate. Right now I feel strongly like we have been tip toeing around really annoying problems for 5 years and, as a result, I at least code workarounds instead of using the factored out, well written code that we should be using from perennial. This may mean copying quite a bit over into perennial/dual/

**Comment by samreid** [comment]
That sounds reasonable to me, and I think we should move updateCopyrightDates et. al. over to perennial.  But my main hesitation on dual/ is that deep down I don't really understand why perennial is not versioned like the rest of our repos.  Perhaps we can discuss that a bit more for my understanding, then create a side issue assigned to me for updateCopyrightDates?

**Comment by zepumph** [comment]
Seems best for a quite refresher at dev meeting, and to PSA about this new pattern.

**Comment by samreid** [comment]
Also, let's have a PSA to exclude chipper/dual.

**Comment by zepumph** [comment]
>@zepumph do you want a git add command to run automatically, or should it be manual for now?


Automatic is best to me.

**Comment by jonathanolson** [comment]
In https://github.com/phetsims/perennial/commit/896bf3e6b95e7d59ff2ed34f433ec3e5b7d97734, it looks like a grunt dependency was added to execute? Will this work nicely when not running at all from grunt? We need it to not print out any lines, and was hoping to not be adding grunt dependencies to things if we someday want to remove grunt.

**Comment by jonathanolson** [comment]
> I don't really understand why perennial is not versioned like the rest of our repos.

That is its purpose, to always be on master (thus the name).

**Comment by samreid** [comment]
Yes, but what problems would it cause for a sim to check out a specific SHA of perennial as part of a build?  If the only problem is that we are running code from perennial, so we don't want to check out code from it while it is running, then could that be solved by the following new proposal?

* Check out perennial to two directories: `git clone perennial perennial-master` and `git clone perennial perennial-versioned`.  We run commands out of perennial-master, and sims can depend on perennial-versioned?  Then we deal with 2 checkouts of perennial instead of the overhead of dual/

**Comment by jonathanolson** [comment]
> Check out perennial to two directories

That sounds possible, mainly like taking dual/ to a new level? It breaks some assumptions that some of our code uses, however, so it would require some changes.

**Comment by samreid** [comment]
It's also not clear what would go wrong if you are running a perennial/master task and it checks out an older version of perennial in place.  If you want to run another task from perennial/master, you just check out master again, right?  Or do we have processes that require the shas on disk while starting up another perennial process?

**Comment by zepumph** [comment]
Wouldn't dividing perennial into two repos suffer from the exact same problem we currently have, because when you checked out an older version of the sim, you would be stuck having master perennial and an older version of `dual` (the repo), and thus an older version of `execute` to be run on master perennial.

**Comment by samreid** [comment]
Chipper would depend on perennial-versioned, not perennial-master.  But I'd also like to discuss why we need 2 perennials, based on the remarks in https://github.com/phetsims/chipper/issues/1018#issuecomment-879297017. Let's hypothesize that (a) `require` files are loaded into memory on startup rather than dynamically, and (b) there are no dynamically required or invoked parts.  Then you can kick off a perennial task that changes perennial's SHAs and it will be OK.

**Comment by jonathanolson** [comment]
> It's also not clear what would go wrong if you are running a perennial/master task and it checks out an older version of perennial in place. If you want to run another task from perennial/master, you just check out master again, right? Or do we have processes that require the shas on disk while starting up another perennial process?

It doesn't sound like a good practice to be changing checkouts while things are running. Perennial should have the ability to run new scripts inside itself.

Having to program things so that you can't depend on the files in your repo changing out from underneath while running sounds like a lot of complexity.

**Comment by zepumph** [comment]
>Chipper would depend on perennial-versioned, not perennial-master.

Yes, but perennial-master would depend on perennial-versioned, so @jonathanolson's worry about files changing out from under perennial still apply. If perennial-master ever wants to use `execute`, it could check out old shas, which means old `execute`, and perennial-master won't necessarily work with old `execute`.

**Comment by jonathanolson** [comment]
> Yes, but perennial-master would depend on perennial-versioned

I'm not understanding this... wouldn't perennial-master just use its own `execute`? Perennial's "master" (i.e. perennial) should not depend on its copy (i.e. perennial-versioned).

**Comment by zepumph** [comment]
Yes sorry I didn't at first understand the proposal. This seems very clear to me now.

>Check out perennial to two directories: git clone perennial perennial-master and git clone perennial perennial-versioned. We run commands out of perennial-master, and sims can depend on perennial-versioned? Then we deal with 2 checkouts of perennial instead of the overhead of dual/

I like this best so far!

**Comment by samreid** [comment]
One more idea (shortcut?) to simplify this proposal and make it easy to start using it.  We would leave "perennial" as it is now, so that nothing about its current usage needs to change. Then we give a new name to the versioned checkout. "perennial-versioned" or something catchy.   Is "annual" the botanical opposite of perennial?

**Comment by jonathanolson** [comment]
> We would leave "perennial" as it is now, so that nothing about its current usage needs to change. Then we give a new name to the versioned checkout.

That sounds best to me, for this path.

**Comment by samreid** [comment]
We mentioned this to the dev-meeting team today, then met further as a dev-team subgroup: @zepumph @jonathanolson and myself.  We considered many aspects and agreed the proposal is a significant improvement over the status quo, and I volunteered to implement the initial draft.

This will mean the following structure changes:
* changing clone commands to clone a second copy of perennial
* adding the second copy of perennial to the sim repo lists
* adding the clone commands to README.md files that list dependencies.
* adding the new repo clone as a dependency

Once this is done, we can instruct developers to:
* clone-missing-repos
* exclude the secondary/versioned checkout from the IDE
* develop in the primary perennial clone
* pull the secondary/versioned checkout when (a) running code in master that is `required` from chipper.  This is the main process change for developers, and is the alternative to "make a change in perennial, then run a script that copies the file to chipper where a copy of the file will also need to be committed"

We can then have chipper point to the secondary checkout for its require() statements.

We considered whether this would need to be approved by the dev team before we start, but agreed that (a) the proposal is significantly better than the existing strategy and (b) if someone else on the dev team has a better idea, we will be happy to hear it!

I'm leaning toward naming the copy "annual" as it is the botanical opposite of "perennial", and has the connotation that it is transient, and the words feel like they are from the same family.

**Comment by samreid** [comment]
Questions as I get started:
* How can we make sure chipper doesn't import any files from perennial?  Lint rule?

**Comment by samreid** [comment]
- [ ] Notify dev team
- [ ] Notify Google group
- [ ] Clean up dual/ files
- [ ] Test with chains

**Comment by samreid** [comment]
I'm struggling to get `npm update` running on annual.  it's missing winston, the build server says:

```
Build failure: Error: Build aborted, Error: grunt --allHTML --debugHTML --brands=phet,phet-io --locales=en in ../chains failed with exit code 3
stdout:
Loading "Gruntfile.js" tasks...ERROR
>> Error: Cannot find module 'winston'
>> Require stack:
>> - /data/share/phet/phet-repos/annual/js/dual/execute.js
>> - /data/share/phet/phet-repos/chipper/js/grunt/copySupplementalPhetioFiles.js
>> - /data/share/phet/phet-repos/chipper/js/grunt/buildRunnable.js
>> - /data/share/phet/phet-repos/chipper/js/grunt/Gruntfile.js
>> - /data/share/phet/phet-repos/chains/Gruntfile.js
>> - /data/share/phet/phet-repos/chains/node_modules/grunt/lib/grunt/task.js
>> - /data/share/phet/phet-repos/chains/node_modules/grunt/lib/grunt.js
>> - /usr/lib/node_modules/grunt-cli/bin/grunt
Warning: Task "default" not found.? Use --force to continue.
```

**Comment by samreid** [comment]
I tried many things but could not get past the missing winston problem.  It would probably have worked to remove winston  from execute, but that's perhaps not a desirable long term solution (unless we have perennial occurrences pass in winston as an option, and have a perennialExecute wrapper).  I'll need to wait for assistance from @zepumph @mattpen or @jonathanolson.  For now I reverted the commits in chipper and perennial, and am confirming a chains build works ok.

**Comment by samreid** [comment]
Another option would be to create a new repo that perennial and chipper depend on, which has no node_modules.  At the moment, this would contain SimVersion and execute.

**Comment by zepumph** [comment]
>Another option would be to create a new repo that perennial and chipper depend on, which has no node_modules. At the moment, this would contain SimVersion and execute. winston would have to be passed in.

This would not work though right, because that new repo could be on an old version while perennial is on master and trying to use something only found on master of the new repo.



@samreid, I believe you would have needed to restart the build server since you changed a script that the server would use (and likely that was kept in memory of the process on phet-server. Did you restart build-server last night?

**Comment by samreid** [comment]
Thanks, I was not aware the build server needed to be restarted.  I had trouble with permissions though, even trying to journalctl following the instructions and using the listed credentials, so I may need help restarting build server as well.

> This would not work though right, because that new repo could be on an old version while perennial is on master and trying to use something only found on master of the new repo.

To make it work, say the new repo is called "perennial-core", then it would be cloned twice, once as "perennial-core" and once as "perennial-core-versioned".  This proposal wouldn't use annual/.  Perennial would depend on perennial-core and chipper would depend on perennial-core-versioned.  I don't want this to get overcomplicated though, perhaps we should test the primary proposal now that I know to restart the build server.

**Comment by zepumph** [comment]
It seems like the primary benefit of that would be that it would be clear what files in perennial were allowed to be used by chipper. Right now I wasn't aware that we were putting any such restrictions on what chipper could use, though there are some obvious ones that shouldn't be (like high level MR, build-server, and deployment tasks). I feel like factoring half of perennial out to another repo could be more confusing, and more trouble than it's worth.

**Comment by samreid** [comment]
@mattpen and I discussed this today, and set an appointment for next week to collaborate on making and testing these changes again.

**Comment by samreid** [comment]
I'm seeing more places that should be using annual/.  For instance, phetmarks, skiffle, rosetta.  phet-info, binder and aqua I'm not sure whether they are "always master" or versioned.  @jonathanolson can you comment on whether aqua should import files from perennial or aqua?

**Comment by zepumph** [comment]
phetmarks should depend on annual, that will be really nice. rosetta is not versioned, binder should stay on master, and I think aqua should also stay on master and use perennial. phet-info seems totally fine to stay on master. I have no ideal what skiffle is.

**Comment by samreid** [comment]
I added a note to work on this Friday Aug 6, unassigning until then.

**Comment by samreid** [comment]
After we have annual, we can work on:
https://github.com/phetsims/phetmarks/issues/54
https://github.com/phetsims/chipper/issues/1064

**Comment by samreid** [comment]
Working with @mattpen, we reverted the revert commits and tested chains rc and production, and a new RC of chains so it would pick up fresh chipper SHAs.  Everything checked out OK.  We coordinated with @zepumph to make sure we got a good merge on daily-grunt-work.sh.  No problems noted so far.  I'll notify slack and the google group, and let's keep alert for problems related to this change.

**Comment by pixelzoom** [comment]
These changes definitely warrant an announcement on the Google Group. Is that on the TODO list?

**Comment by pixelzoom** [comment]
- [ ] All of the sim README.md files need to be updated to include cloning of annual.  

- [ ] PhET Development overview example needs to be updated to include cloning of annual.

**Comment by zepumph** [comment]
Don't forget about the TODOs in https://github.com/phetsims/chipper/issues/1018#issuecomment-893648510

**Comment by chrisklus** [comment]
From 8/5/21 dev meeting:

Since perennial and annual should be kept in sync, we would like to investigate a post-push hook to pull annual if it's on the same branch as perennial. 

MK: post-push hooks don't exist. We will keep thinking about it. pre-push vocal reminder to pull annual?

SR: Glad we made devs aware, hopefully someone will come up with an idea for the best workflow when encountering this naturally.

CM: any strong feelings about the name of the perennial clone? 
MK: If we switched to only having the files from perennial that need to be versioned in annual, and perennial and chipper depended on that, we would like the repo to be named annual.
JB: I would like something like `perennial-versioned` since `annual` makes it seem like it's an independent repo, but it doesn't actually exist at `github/phetsims`.
MK: I'm excited about `perennial-versioned`

SR: Preference for `annual`, happy to be outvoted. Seems like no one else has this preference. We will go with `perennial-versioned`

**Comment by chrisklus** [comment]
We changed our mind to `perennial-alias`

**Comment by zepumph** [comment]
- [ ] Don't forget that we likely don't need `winston` in chipper anymore.

**Comment by samreid** [comment]
@pixelzoom recommendations from https://github.com/phetsims/chipper/issues/1018#issuecomment-893648510 got buried by commits, so I'll copy them here:

Loose ends:

- [ ] Make an annoucement about these breaking changes in the Google Group.
- [ ] README.md files need to be regenerated to include `git clone https://github.com/phetsims/perennial.git`
- [ ] [PhET Development Overview](https://github.com/phetsims/phet-info/blob/master/doc/phet-development-overview.md) "Getting Started" sections need to be updated to include cloning of annual.
- [ ] [PhET Development Overview](https://github.com/phetsims/phet-info/blob/master/doc/phet-development-overview.md) "Creating a New Sim" section needs to be updated to include `npm install` for annual.
- [ ] [Deployment info for chipper 2.0](https://github.com/phetsims/phet-info/blob/master/deployment-info/chipper-2.0.md#manual-maintenance-patching) "Deploying simulation" section talks about `npm update` for perennial. Does it need to mention annual now?
- [ ] `grunt --help` in chipper shows "Run grunt --help in perennial to see documentation" for many of the options. Is that still correct, or should it be pointing to annual? 
- [ ] Because perennial and annual are the same git repo, local copies that are on the same branch should be kept in sync. This responsibility currently falls on the developer to handle manually. So any checklist or doc that says "push to perennial" should also include "pull annual". E.g. new_repo_checklist.md, sim_checklist.md, updating_metadata.md.

Also from the other TODOs:
- [ ] Update dual

Also, I'd like to mention https://github.com/phetsims/perennial/issues/235 where there is a proposal that may alleviate a lot of these problems.

**Comment by samreid** [comment]
@pixelzoom recommendations from https://github.com/phetsims/chipper/issues/1018#issuecomment-893648510 got buried by commits, so I'll copy them here:

Loose ends:

- [ ] Make an annoucement about these breaking changes in the Google Group.
- [ ] README.md files need to be regenerated to include `git clone https://github.com/phetsims/perennial.git`
- [ ] [PhET Development Overview](https://github.com/phetsims/phet-info/blob/master/doc/phet-development-overview.md) "Getting Started" sections need to be updated to include cloning of annual.
- [ ] [PhET Development Overview](https://github.com/phetsims/phet-info/blob/master/doc/phet-development-overview.md) "Creating a New Sim" section needs to be updated to include `npm install` for annual.
- [ ] [Deployment info for chipper 2.0](https://github.com/phetsims/phet-info/blob/master/deployment-info/chipper-2.0.md#manual-maintenance-patching) "Deploying simulation" section talks about `npm update` for perennial. Does it need to mention annual now?
- [ ] `grunt --help` in chipper shows "Run grunt --help in perennial to see documentation" for many of the options. Is that still correct, or should it be pointing to annual? 
- [ ] Because perennial and annual are the same git repo, local copies that are on the same branch should be kept in sync. This responsibility currently falls on the developer to handle manually. So any checklist or doc that says "push to perennial" should also include "pull annual". E.g. new_repo_checklist.md, sim_checklist.md, updating_metadata.md.

Also from the other TODOs:
- [ ] Update dual

Also, I'd like to mention https://github.com/phetsims/perennial/issues/235 where there is a proposal that may alleviate a lot of our problems (at the cost of one potentially easier-to-deal-with problem).

**Comment by mattpen** [comment]
@samreid - there was a build failure last night due to the perennial-alias repo.  This was during an `ar` translation of build-an-atom 1.6.16.

`Aug 05 23:35:48 phet-server.int.colorado.edu build-server[11555]: `

```
debug: Running grunt --brands=phet --repo=build-an-atom build-images from ../chipper
debug: Command grunt finished. Output is below.
debug: stderr is empty.
debug: stdout: Loading "Gruntfile.js" tasks...ERROR
>> Error: Cannot find module 'winston'
>> Require stack:
>> - /data/share/phet/phet-repos/perennial-alias/js/dual/execute.js
>> - /data/share/phet/phet-repos/chipper/js/grunt/copySupplementalPhetioFiles.js
>> - /data/share/phet/phet-repos/chipper/js/grunt/buildRunnable.js
>> - /data/share/phet/phet-repos/chipper/js/grunt/Gruntfile.js
>> - /data/share/phet/phet-repos/chipper/Gruntfile.js
>> - /data/share/phet/phet-repos/chipper/node_modules/grunt/lib/grunt/task.js
>> - /data/share/phet/phet-repos/chipper/node_modules/grunt/lib/grunt.js
>> - /usr/lib/node_modules/grunt-cli/bin/grunt
Warning: Task "build-images" not found.^G Use --force to continue.
Aborted due to warnings.
```

Should we check out eb080df17e7d77102f410d5d036a23e12c48172e and restart the build server?

**Comment by samreid** [comment]
@mattpen can you speculate as to why /data/share/phet/phet-repos/perennial-alias/ is missing winston?  Would manually running `npm install` in that directory help?

**Comment by mattpen** [comment]
I ran `sudo -u phet-admin npm i` in `/data/share/phet/phet-repos/perennial-alias`, then submitted an arabic translation of chains.  The build worked without error.

**Comment by zepumph** [comment]
But the real issue is to find where In the build server process we need to add a call to npm update perennial-alias. Until the  this is just a bandaid until another dependency is added for files used by chipper in -alias. That is the root issue here.

**Comment by zepumph** [comment]
- [ ] We need to reupdate the clone commands, see https://github.com/phetsims/chipper/blob/d9391a46ecd9174c6556af6e3125774266df85b9/js/grunt/generateREADME.js#L30-L36 for outdated usages of "annual"

**Comment by zepumph** [comment]
Oh shoot! I deleted the above command because I thought I had pulled fully in chipper, but I had not. Nevermind.

**Comment by samreid** [comment]
At a staggering 512 hidden items, this issue is ready to split into sub-issues.

**Comment by samreid** [comment]
I opened side issues for the remaining work.  Anyone please feel free to reopen if something was missed.  Closing.

**Comment by jonathanolson** [comment]
Reopening, since old checkouts are trying to access perennial. Will need to maintenance release to fix.

**Comment by samreid** [comment]
Sounds like @jonathanolson is on the case for this maintenance fix, so I've updated the assignees.

**Comment by jonathanolson** [comment]
Handled in the maintenance release, closing.
